src/Avalonia.Base/Media/Imaging/PixelFormatReaders.cs

```diff
@@ -0,0 +1,280 @@
+        public Rgba8888Pixel ReadNext()
+        {
+            var value16 = *_address;
+            _address++;
+            var value8 = (byte)(value16 >> 8);
+            return new Rgba8888Pixel
+            {
+                A = 255,
+                B = value8,
+                G = value8,
+                R = value8
+            };
+        }
+
+        public void Reset(IntPtr address) => _address = (ushort*)address;
+    }
+
+    public unsafe struct Gray32FloatPixelReader : IPixelFormatReader
+    {
+        private byte* _address;
+        public Rgba8888Pixel ReadNext()
+        {
+            var f = *(float*)_address;
+            var srgb = Math.Pow(f, 1 / 2.2);
+            var value = (byte)(srgb * 255);
+
+            _address += 4;
+            return new Rgba8888Pixel
+            {
+                A = 255,
+                B = value,
+                G = value,
+                R = value
+            };
+        }
+
+        public void Reset(IntPtr address) => _address = (byte*)address;
+    }
+
+    struct Rgba64
+    {
+#pragma warning disable CS0649
+        public ushort R;
+        public ushort G;
+        public ushort B;
+        public ushort A;
+#pragma warning restore CS0649
+    }
+
+    public unsafe struct Rgba64PixelFormatReader : IPixelFormatReader
+    {
+        private Rgba64* _address;
+        public Rgba8888Pixel ReadNext()
+        {
+            var value = *_address;
+
+            _address++;
+            return new Rgba8888Pixel
+            {
+                A = (byte)(value.A >> 8),
+                B = (byte)(value.B >> 8),
+                G = (byte)(value.G >> 8),
+                R = (byte)(value.R >> 8),
+            };
+        }
+
+        public void Reset(IntPtr address) => _address = (Rgba64*)address;
+    }
+
+    public static void Transcode(IntPtr dst, IntPtr src, PixelSize size, int strideSrc, int strideDst,
+        PixelFormat format)
+    {
+        if (format == PixelFormats.BlackWhite)
+            Transcode<BlackWhitePixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray2)
+            Transcode<Gray2PixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray4)
+            Transcode<Gray4PixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray8)
+            Transcode<Gray8PixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray16)
+            Transcode<Gray16PixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray32Float)
+            Transcode<Gray32FloatPixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Rgba64)
+            Transcode<Rgba64PixelFormatReader>(dst, src, size, strideSrc, strideDst);
+        else
+            throw new NotSupportedException($"Pixel format {format} is not supported");
```

Comment:
From the list of these formats, seems like Rgb565 is missing as D2D backend doesn't support it (as I see from this PR). Low priority, I suppose.